src/main/java/org/mule/extension/http/internal/request/HttpPollingSource.java

```diff
@@ -181,53 +179,55 @@ private String getResolvedUri() {
+    }
+  }
+
   private void sendRequest(PollContext<String, HttpResponseAttributes> pollContext) {
-    CompletionCallback<InputStream, HttpResponseAttributes> callback =
-        new CompletionCallback<InputStream, HttpResponseAttributes>() {
-
-          @Override
-          public void success(Result<InputStream, HttpResponseAttributes> result) {
-            HttpResponseAttributes attributes = result.getAttributes().orElse(null);
-            MediaType mediaType = result.getMediaType().orElse(ANY);
-            Reference<Boolean> atLeastOneResult = new Reference<>(false);
-            getItems(result.getOutput(), mediaType, attributes).forEach(item -> {
-              atLeastOneResult.set(true);
-              pollContext.accept(getPollingItemConsumer(item));
-            });
-
-            if (!atLeastOneResult.get()) {
-              LOGGER.debug("Empty result in HTTP Polling Source at {} of uri {}", location.getRootContainerName(), resolvedUri);
-            }
-          }
-
-          @Override
-          public void error(Throwable throwable) {
-            LOGGER.error("There was an error in HTTP Polling Source at {} of uri '{}'", location.getRootContainerName(),
-                         resolvedUri, throwable);
-          }
-        };
+    Serializable currentWatermark = pollContext.getWatermark().orElse(null);
+    requestBuilder.updateWatermark(currentWatermark);
 
     LOGGER.debug("Sending '{}' request to '{}' in flow '{}'.", method, resolvedUri, location.getRootContainerName());
     try {
-      httpRequester.doRequest(client, config, resolvedUri, method, config.getRequestStreamingMode(), config.getSendBodyMode(),
-                              config.getFollowRedirects(), client.getDefaultAuthentication(), config.getResponseTimeout(),
-                              getResponseValidator(), transformationService, requestBuilder, true, muleContext, scheduler, null,
-                              null, callback, injectedHeaders, null);
-    } catch (MuleRuntimeException e) {
-      LOGGER.error("Trigger '{}': Mule runtime exception found while executing poll to {}: '{}'", getId(), resolvedUri,
-                   e.getMessage(), e);
+      Result<InputStream, HttpResponseAttributes> result =
+          httpRequester.doSyncRequest(client, config, resolvedUri, method, config.getRequestStreamingMode(),
+                                      config.getSendBodyMode(), config.getFollowRedirects(), client.getDefaultAuthentication(),
+                                      config.getResponseTimeout(), getResponseValidator(), transformationService, requestBuilder,
+                                      true, muleContext, scheduler, injectedHeaders)
+              .get();
+      pollResult(pollContext, result, currentWatermark);
+    } catch (ExecutionException | InterruptedException e) {
+      LOGGER.error("There was an error in HTTP Polling Source at {} of uri '{}'", location.getRootContainerName(), resolvedUri,
```

Comment:
the interrupted exeption is not being properly handled here. Ref: https://www.baeldung.com/java-interrupted-exception